tune_sklearn/_trainable.py

```diff
@@ -314,3 +320,102 @@ def reset_config(self, new_config):
+                self.estimator_config[
+                    f"{self.main_estimator.steps[-1][0]}__warm_start"] = True
+                self.estimator_config[
+                    f"{self.main_estimator.steps[-1][0]}__max_iter"] = 1
+
+            elif self._can_warm_start_ensemble():
+                # Each additional call on a warm start ensemble only trains
+                # new estimators added to the ensemble. We start with 0
+                # and add an estimator before each call to fit in _train(),
+                # training the ensemble incrementally.
+                self.estimator_config[
+                    f"{self.main_estimator.steps[-1][0]}__warm_start"] = True
+                self.estimator_config[
+                    f"{self.main_estimator.steps[-1][0]}__n_estimators"] = 0
+
+        return n_splits
+
+    def _early_stopping_partial_fit(self, i, estimator, X_train, y_train):
+        """Handles early stopping on estimators that support `partial_fit`.
+
+        """
+        if hasattr(estimator, "partial_fit"):
+            estimator.partial_fit(X_train, y_train, np.unique(self.y))
+        else:
+            # Workaround if the pipeline itself doesn't support partial_fit
+            # (default sklearn doesn't). We set the final step to "passthrough",
+            # so that it doesn't get executed, do fit_transform to get transformed
+            # data, and then call partial_fit on just the final step.
+            last_step = estimator.steps[-1][1]
+            estimator.steps[-1] = (estimator.steps[-1][0], "passthrough")
+            X_train_transformed = estimator.fit_transform(X_train, y_train)
+            estimator.steps[-1] = (estimator.steps[-1][0], last_step)
+            estimator.steps[-1][1].partial_fit(X_train_transformed, y_train,
+                                               np.unique(self.y))
+
+    def _early_stopping_xgb(self, i, estimator, X_train, y_train):
+        """Handles early stopping on XGBoost estimators.
+
+        """
+        estimator.fit(X_train, y_train, {
+            f"{self.main_estimator.steps[-1][0]}__xgb_model": self.
+            saved_models[i]
+        })
+        self.saved_models[i] = estimator.get_booster()
+
+    def _early_stopping_iter(self, i, estimator, X_train, y_train):
+        """Handles early stopping on estimators supporting `warm_start`.
+
+        """
+        estimator.fit(X_train, y_train)
+
+    def _early_stopping_ensemble(self, i, estimator, X_train, y_train):
+        """Handles early stopping on ensemble estimators.
+
+        """
+        # User will not be able to fine tune the n_estimators
+        # parameter using ensemble early stopping
+        updated_n_estimators = estimator.get_params(
+        )[f"{self.main_estimator.steps[-1][0]}__n_estimators"] + 1
+        estimator.set_params(
+            **{
+                f"{self.main_estimator.steps[-1][0]}__n_estimators": updated_n_estimators
+            })
```

Comment:
for _PipelineTrainable, how about let's do:

```
@property
def main_estimator_name(self):
    return self.main_estimator.steps[-1][0]

Reply:
so then this becomes:

```suggestion
        n_estimator_key = f"{self.main_estimator_name}__n_estimators"
        updated_n_estimators = estimator.get_params()[n_estimator_key] + 1
        estimator.set_params(
            **{n_estimator_key: updated_n_estimators})
```

Reply:
Yep, looks much better that way, thanks